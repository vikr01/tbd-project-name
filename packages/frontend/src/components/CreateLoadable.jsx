// @flow
import React from 'react';
import Loadable from 'react-loadable';
import type { Node } from 'react';
import { Button } from '@material-ui/core';

type LoadingProps = {
  error: boolean,
  retry: Function,
};

const defaultLoading = ({ error, retry }: LoadingProps): Node => {
  if (error) {
    return (
      <div>
        Error!
        <Button onClick={retry}>Retry</Button>
      </div>
    );
  }
  return <div>Loading...</div>;
};

type Props = {
  loader: Node,
  children: ?Node,
};

const CreateLoadable = ({
  loader,
  children: loading = defaultLoading,
  ...otherProps
}: Props): Node => {
  const LoadableComponent = Loadable({
    loader,
    loading,
    ...otherProps,
  });

  return <LoadableComponent />;
};

export default CreateLoadable;
