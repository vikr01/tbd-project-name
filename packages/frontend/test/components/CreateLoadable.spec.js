import React from 'react';
import renderer from 'react-test-renderer';
import delay from 'delay';
import CreateLoadable from 'tbd-frontend-name/src/components/CreateLoadable';

describe('CreateLoadable', () => {
  test('works with a simple component', async () => {
    const createdLoadable = (
      <CreateLoadable loader={async () => () => <div>Foo</div>} />
    );
    const tree = renderer.create(createdLoadable);

    expect(tree.toJSON()).toMatchSnapshot();

    await delay(100);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('handles errors', async () => {
    const createdLoadable = (
      <CreateLoadable
        loader={async () => {
          throw new Error('foo');
        }}
      />
    );

    const tree = renderer.create(createdLoadable);

    expect(tree.toJSON()).toMatchSnapshot();

    await delay(100);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
