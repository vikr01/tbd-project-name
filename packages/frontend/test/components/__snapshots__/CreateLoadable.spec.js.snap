// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateLoadable handles errors 1`] = `
<div>
  Loading...
</div>
`;

exports[`CreateLoadable handles errors 2`] = `
<div>
  Error!
  <button
    className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-2"
    >
      Retry
    </span>
    <span
      className="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;

exports[`CreateLoadable works with a simple component 1`] = `
<div>
  Loading...
</div>
`;

exports[`CreateLoadable works with a simple component 2`] = `
<div>
  Foo
</div>
`;
